### ISSUE TITLE
Asset and data set creation forms should only show one input options at a time

### ISSUE THREAD

**Issue by ghing** [denverfoundation/storybase#767]
Originally reported by @bpawlak in [a comment](https://github.com/PitonFoundation/atlas/issues/583#issuecomment-18700725) to #583 

> I know we're using the "do this OR do that" for things like uploading an image (e.g., enter 
> URL or upload image) but when you get to 3 options like on adding a chart, it's confusing. 
> And the message "You must specify only one of the following values body, url" (if you try 
> to enter a chart URL and Embed code) isn't very clear to users. What about something 
> similar to using radio buttons to expose the field such that only 1 is shown at a time:

![Asset content selection mockup](https://f.cloud.github.com/assets/2250519/586991/15f524cc-c95b-11e2-8100-4385d4cdc9c5.png)

**Comment by ghing** [comment]
Updated title to reflect that this is also an issue with data sets.

**Comment by ghing** [comment]
@bpawlak I prefer tabs or pills to the radio buttons. Do you have a strong preference between:
- radio buttons
- tabs
- pills

**Comment by bpawlak** [comment]
If we use tabs/pills, then they're laid out horizontally?  We don't use pills anywhere else do we?

**Comment by ghing** [comment]
@bpawlak, the tabs/pills would be laid out horizontally, maybe like this:

![2013-06-18-122604](https://f.cloud.github.com/assets/110420/670101/3af6b548-d83c-11e2-957e-a659c58ec99e.jpg)

We use pills for the featured image:

![screenshot from 2013-06-18 12 18 23](https://f.cloud.github.com/assets/110420/670106/468ef1fe-d83c-11e2-9ac7-432bdcb296ba.png)

**Comment by bpawlak** [comment]
Nice "screenshot!"  :)

I forgot about the tabs in the featured image.  If Jordan's ok with it, then we should do that - keep consistency in the app

**Comment by ghing** [comment]
@patternleaf, so the implementation for this is just about showing/hiding certain form fields on the front-end.  The elements don't have to be removed from the DOM (unless that's easier) and the back-end already properly ignores empty fields/values.

It might be helpful to look at the documentation and the source code for [Backbone Forms](https://github.com/powmedia/backbone-forms) which is what's being used for asset creation forms. It might make sense to make a subclass of `Backbone.Form` that adds the additional methods for toggling the fields.  Also note that `Backbone.Form` is just a subclass of `Backbone.View`, so you can use the patterns you're used to with views with a form.

Note that we only have to worry about this for the form used when creating a new asset/dataset.  `Asset.schema` and `DataSet.schema` handle showing only the relevant fields for the form that has already been created.

One thing I couldn't decide on when I first looked at this issue was the best, most semantic markup for the different content fields.

**Comment by patternleaf** [comment]
@ghing How about fieldsets? Looks like we're already using one at the top level of the form, but apparently it's ok to nest them.

On the other hand we're currently wrapping each field in an `li` which isn't terrible I suppose.

**Comment by ghing** [comment]
@patternleaf, I was leaning toward fieldsets too, just wasn't sure the best way to make use of them.

**Comment by bpawlak** [comment]
Do we need to keep an eye out for non-blank forms that aren't currently displayed in the UI?  That is, a user can start by uploading a file and then specifying a URL

**Comment by ghing** [comment]
@bpawlak, no.  Once a user either uploads or specifies a URL or pastes an embed code, the asset storage mechanism is locked in and they'll only ever see the fields for that type of content.  So, we're really only concerned about disambiguating the fields when the asset is initially created.

**Comment by bpawlak** [comment]
So whichever "panel" is viewable when they submit the form is the one that gets submitted?

**Comment by patternleaf** [comment]
@ghing is it difficult to change that behavior on the back-end? off-hand i'm finding it confusing to, say, upload an image, then click Edit, then see a new blank upload form without other options.

@ghing @bpawlak @jwirfs-brock There is an implementation of this—for asset creation only—currently on http://dev2.floodlightproject.org

**Comment by ghing** [comment]
@bpawlak, I misunderstood what you were saying.  Currently, and with the new design, they'll get an error message if they fill out multiple content fields (e.g. URL and Upload).  So, we just need to make sure that the error message is in a visible and clear place.  Alternately, we could disable the tab toggling once a value is set for a particular content field.

**Comment by ghing** [comment]
@patternleaf, it would be pretty hard to change this on the back end.  I think it's most weird with the image upload because that's the one where you just see a blank upload form.  For the other asset types, it's not as weird.  I think we could make the the image editing more clear if we added a descriptive help string.

The whole intention with the design is that an asset is a specific piece of content with metadata.  Not "content in this space".  So we allow people to edit it because they might want to fix a typo in a URL, or update a caption or change the width parameter of an embed URL.  But if they actually want to change the content (e.g. point to a different image), they should remove the asset and create a new one.  Doing this also makes the old content available in the asset drawer if they want to re-use it later.

**Comment by bpawlak** [comment]
I would suggest we find a way to only submit what's viewable to the user... It's a common problem with tab-based form inputs and users generally only expect that the final selected tab is the one that's submitted.

**Comment by ghing** [comment]
@patternleaf I took a look at what's currently up on dev2.  This works as I'd imagined, but:
- The stacked pills look weird (I think this is only on narrow, side-by-side layouts)
- The long "Or, select an image from your own computer to be included in your story." label sucks.  I'm thinking that there should be an overarching title like "[ASSET_TYPE] Content", then sub labels for each of the pills, e.g. "As a URL", "Upload an image", "Paste an embed code" and then a label for the actual HTML form input/textarea.  Thoughts?
- After your [previous comment](https://github.com/PitonFoundation/atlas/issues/767#issuecomment-20813345) I think we should have different help text for the upload field when editing an uploaded asset that specifies that the upload field replaces the existing image and suggesting that the user might want to remove the asset instead ("You can replace the image with a new file from your computer.  If you think you'll want to use this image later, it's better to remove this asset and create a new one with the preferred image").  Or we could remove the upload field altogether and just have a message like "If you'd like to replace this image, remove the asset and create a new one. This asset will be available in the asset drawer if you'd like to use it later".

Finally, when you're finishing this implementation, can you create a test script like [this one for datasets](https://github.com/PitonFoundation/atlas/blob/develop/apps/storybase_asset/features/asset_datasets.feature) so we can make sure we cover all the different states with this. I think there are probably a lot.

**Comment by ghing** [comment]
@bpawlak, re [your comment](https://github.com/PitonFoundation/atlas/issues/767#issuecomment-20815450)

> I would suggest we find a way to only submit what's viewable to the user... 

That makes sense.  

@patternleaf, hit me up if you need ideas about how to do this.  Off the top of my head, you could have a `submit` event handler that cleans the fields in the other tabs, or you could have a hidden form field that passes the visible "tab" and handle the cleaning in the Form's validation method.

**Comment by patternleaf** [comment]
@ghing @bpawlak @jwirfs-brock @tekhneco Wanted to solicit feedback on a couple of things.
1. I added a title to the asset edit/creation form.
2. Organized the tabbed interface inside a fieldset container with legend, with a border and background, etc.

Screenshot:

![screenshot 7 15 13 10 47 am](https://f.cloud.github.com/assets/73439/799132/7f44bd98-ed6e-11e2-9181-aa07f103e737.jpeg)

Some rationale: I like the fieldset with legend as a way to organize the tabs (so the pills aren't just floating out in space), but I felt it made things way too boxes-within-boxes-within-boxes. I added the light gray background in hopes of setting off that box visually. And it's arguably the most important part of the form anyway.

For the title, there was actually already a header element there but nothing was being rendered into it. I imagine this was just an oversight at some point.

@ghing, FYI I couldn't see a way to use backbone-form templates to good effect here. Its templates are organized by individual field, and this little control is all about _combining_ fields ... if you have other thoughts, of course let me know.

**Comment by patternleaf** [comment]
Maybe the title should be "Add [Asset Type]" rather than "Create" ...

**Comment by bpawlak** [comment]
The pills "floating in space" is always an issue with rendering tabs as pills and one reason why I don't really like them, especially for infrequent usage.  Although I like the idea of putting them into a container, the grey background somehow feels like this container is required or selected or something.  Do we use fieldsets anywhere else in the application?

Agree on "Add Asset" instead of "Create"

**Comment by jwirfs-brock** [comment]
This works the same way that the select featured image does on the Publish/Share step, right? I like it -- at least, I like it as it appears there. It feels very intuitive to me.

I also agree that "Add Chart"/"Add Image", etc. works better than "Create..."

Is the screen on the right what it would look like after you've already selected an input mechanism and then saved it? Is the way that it works that once you've done that (i.e., once you've added an asset via a URL), you can't go back and decide you want to upload that asset as a file instead, right? Just wanting to make sure I understand this correctly...

**Comment by patternleaf** [comment]
@bpawlak 

>  Do we use fieldsets anywhere else in the application?

We use the fieldset and legend tags elsewhere (http://floodlightproject.org/en/accounts/notifications/, for example), but I don't believe it's being rendered in this way elsewhere.

> the grey background somehow feels like this container is required or selected or something

Yeah ... I'm not thrilled with this either. Here's a shot without the gray:

![screenshot 7 15 13 11 42 am](https://f.cloud.github.com/assets/73439/799498/11f447ce-ed76-11e2-89f0-83f8a851ffde.jpeg)

Without any decoration/container, and a hidden legend:

![screenshot 7 15 13 11 45 am](https://f.cloud.github.com/assets/73439/799509/694aac5c-ed76-11e2-80e1-b3da231e6b6d.jpeg)

And some mildly tabified pills:

![screenshot 7 15 13 11 46 am](https://f.cloud.github.com/assets/73439/799518/939c0b4a-ed76-11e2-97ff-9443aee9a0e1.jpeg)

Now that I'm looking at it again, the decoration-free one actually looks ok to me.

@jwirfs-brock 

> This works the same way that the select featured image does on the Publish/Share step, right?

Yes. It's up and running on dev2 at the moment.

> Is the screen on the right what it would look like after you've already selected an input mechanism and then saved it? Is the way that it works that once you've done that (i.e., once you've added an asset via a URL), you can't go back and decide you want to upload that asset as a file instead, right? Just wanting to make sure I understand this correctly...

Yes, that's right. Geoff [suggested above](https://github.com/PitonFoundation/atlas/issues/767#issuecomment-20815834) adding some help text that explains this.

**Comment by jwirfs-brock** [comment]
@patternleaf In your screenshots above, I prefer the first one (grey border, no fill). It serves to highlight that that is the most important thing you need to fill out (and, in fact, the only required thing). The tab version looks odd to me, like the tabs are controlling everything in the window, not just the source input.

**Comment by bpawlak** [comment]
The orange color of the tab makes it look like a primary button.  I was trying other ways of showing that the pill is actually a control... maybe a small "connector" graphic (again, in a different color than primary-button-orange

![pills](https://f.cloud.github.com/assets/2250519/799605/5b116a3e-ed78-11e2-86f0-5322f88af983.png)

**Comment by patternleaf** [comment]
@bpawlak I like the connector graphic. It certainly shows up in plenty of other places in the builder.

**Comment by patternleaf** [comment]
@jwirfs-brock I personally think the bordered boxes-within-boxes starts to feel noisy. We might be able to treat it somehow to reduce the noise—lightening some borders, say. But I'm kind of a fan of @bpawlak's idea with the arrowed-pill. What did you think of that?

**Comment by jwirfs-brock** [comment]
@bpawlak @patternleaf I don't like the arrowed-pill as much, but it definitely works, so if you two prefer it, let's go that route.

Should we also change this on the featured image selector on the publish/share step?

**Comment by ghing** [comment]
@patternleaf re this:

> FYI I couldn't see a way to use backbone-form templates to good effect here. Its templates are organized by 
> individual field, and this little control is all about combining fields ... if you have other thoughts, of course let me 
> know.

While it's by no means necessary, check out the "[100% custom forms](https://github.com/powmedia/backbone-forms#100-custom-forms)" example from the Backbone Forms docs.  This shows how you use a template for the form at large (and not just templatize the individual fields, or "editors" as Backbone Forms calls them).

**Comment by patternleaf** [comment]
@ghing I did see that; unless I'm missing something I don't think it helps. It would seem awkward to try to fit this particular usage—where some fields want to be grouped and others don't, yet they must still all be dynamically determined by model schema—into the BB.Forms template system. Or rather, we'd still need some DOM-juggling JS run after the form has rendered, so there doesn't seem to be a big win.

@jwirfs-brock:

> Should we also change this on the featured image selector on the publish/share step?

I would say so.

**Comment by patternleaf** [comment]
OK, if anyone (@ghing, @bpawlak, @tekhneco, @jwirfs-brock) is available to do a little extra testing on this (on dev2), that would be great.

In addition to the new form element:

![screenshot 7 29 13 12 33 pm](https://f.cloud.github.com/assets/73439/873688/5b6f472a-f87d-11e2-8396-f20bc34fa201.jpeg)

... there were some oddities/inconsistencies in validation UI. So testing around submitting forms with incomplete/bad data would be helpful. The only changes I made were to client-side validation behavior. So any validation requiring a round-trip to the server is not worked into the new UI.

Other known issues:
- IE8 needs some attention — best to avoid at the moment.
- On Chrome, the "required" attribute triggers a Chrome-supplied validation UI, but no app-native UI. This shows up in the Dataset Name field for example.
- Inconsistencies between dataset form behavior and asset form behavior. For example, in the dataset creation form, leaving out both the URL and File fields results in a message on both fields. In the Image asset creation form, leaving out both fields results in a single message at the top of the form.

**Comment by ghing** [comment]
Regarding your notes above:

> On Chrome, the "required" attribute triggers a Chrome-supplied validation UI, but no app-native UI. This shows up in the Dataset Name field for example.

Browser validation is probably also supported in recent versions of Firefox as well.

@patternleaf and I decided to disable browser validation for these forms because:
- Many forms don't use it
- We can't use it to validate all fields everywhere (even within the asset/dataset creation forms) because some of our validation is based on the values of multiple 
- If we don't use it anywhere, it creates an inconsistent experience with per-field error/warning messages

> Inconsistencies between dataset form behavior and asset form behavior. For example, in the dataset creation form, leaving out both the URL and File fields results in a message on both fields. In the Image asset creation form, leaving out both fields results in a single message at the top of the form.

@patternleaf and I decided to use per-field error messages for fields where there is an error with only that field and an error message above the form when a form fails to validate based on a combination of fields.

Action items:
- [ ] Remove browser validation via `required` attribute on input elements
- [ ] Use single message at top of form for validation errors across multiple fields
- [ ] Different per-field help microcopy based on whether an asset/dataset is being created or updated (based on https://github.com/PitonFoundation/atlas/issues/767#issuecomment-20815834)
- [ ] Test script documenting items that @patternleaf has already tested/thought about testing

**Comment by wendynorris** [comment]
Do you still need form validation testing help? Or did disabling the
browser fix the issue?

**
_wendy norris http://about.me/wendynorris_
ceo + founder :: tekhne media, inc
a digital information and media collective
720 281 9378 :: mobile

*\* http://about.me/wendynorris
*
*

**Comment by patternleaf** [comment]
@tekhneco Testing would still be helpful.

**Comment by ghing** [comment]
@patternleaf, here are some non-bug observations I've made when testing this feature
- [ ] When editing an image, perhaps the field label should be "Replace image" (this may have already been changed)
- [ ] Error message when no fields are present is "You must specify at least one option: url, image." Is this too ambiguous?
- [ ] "Paste embed code for the chart" wraps in the side-by-side layout
- [ ]  Layout of thumbnail looks weird in the side-by-side layout
- [ ] Dataset form content "tabs" microcopy is a little less verbose.  E.g. "Data URL" and "Data file" vs. "Enter table URL".  Should the dataset form labels have verbs in them too? E.g. "Enter data URL" instead of "Data URL"

**Comment by patternleaf** [comment]
Thanks @ghing. 
1. Yes: this should already be changed. 
2. Agreed—this was previous behavior and I didn't get around to changing it. Can fix though.
3. Simple solution: use `white-space:nowrap` so that the end clips instead of wrapping. Do you have another thought? On my screen this only happens in side-by-side layout when the asset or help drawer is also open (or the browser window is narrow, of course).
4. Likewise re: screen width on this one. I'll see what I can do. Apparently I didn't test the UI changes with a narrow enough browser window.

**Comment by patternleaf** [comment]
Found this, too:
1. Load a story with a section with an image asset.
2. Remove the image asset.
3. Select new "Chart" asset type for that slot.
4. Click "Cancel" to not add the chart.
5. Drag and drop the original image asset back into place.
6. Click "Edit Image".

Expected: image editing form shows up.
Actual: "Edit Image" title appears, but with _chart asset Add form fields displayed_.

I imagine this was a pre-existing behavior. I'll take a look.

**Comment by ghing** [comment]
@patternleaf 

Regarding thumbnail, layouts and browser sizes, I'm testing on my notebook that has a 1280x800 resolution.  

I took a quick look at the Google Analytics for screen resolutions (note that I couldn't figure out how to do this specifically for the builder URL), but this is the breakdown for the top recent resolutions site-wide:

| Screen Resolution | Visits |
| --- | --- |
| 1366x768 | 168 |
| 1280x800 | 113 |
| 1440x900 | 102 |
| 1920x1080 | 71 |
| 1280x1024 | 61 |
| 768x1024 | 53 |
| 1024x768 | 51 |
| 1680x1050 | 38 |
| 1600x900 | 34 |
| 320x480 | 30 |

I'm not quite sure how to interpret this, but it seems like we should make sure we test with at least a 1280 wide resolution.

**Update:** I also added screen resolution field to the test matrix spreadsheet so we can see what resolutions the team is testing at and know if we need to test some others.

**Comment by patternleaf** [comment]
Hey @ghing—a case i ran into which we might want to fold in here is when i put in a valid URL to a non-existent file. that is, if we get a 404 or other bad response on a URL, it would be helpful to notify the user somehow. Right now, there's just a blank screen. What do you think?

**Comment by ghing** [comment]
@patternleaf Yeah, I've thought of that before but couldn't figure out the best way to handle it that didn't feel overly expensive.  Do we send a separate request for each URL asset to check if the URL is reachable?  Do we do this client or server side?  I'd open this as a separate issue.

**Comment by patternleaf** [comment]
@ghing Which is another thing: if there is server-side validation for a form input it would be nice to present it in the same way as client-side validation. 

Anyway, in the case of URL-based images, it seems like we should be able to check for bad response codes client side without too much effort. I'll make a new issue.

**Comment by jwirfs-brock** [comment]
@ghing @patternleaf So far, I haven't noticed any errors in the testing. However, I did spot a few non-critical things:
- A user can upload any type of file as a data source file. I tried it with a jpeg and pdf, and both were "allowed." We don't have any instructional copy about the file types we support. Should we? Or should we just leave it completely open? If we do leave it open, I think we should at least include some (very brief) copy about the type of files we recommend (CSV or Excel).
- Now that we have the "Data Sources" feature, and it's working nicely, some of the fields on the asset information view are redundant. Namely, the Table asset type has "Attribution" and "Source URL." I know we originally had those because tables (and charts, and maps), can sometimes be attributed to a source that is _different_ than the data source. But I think that's a subtlety that will, at this point, just end up confusing users. If people really want to source them differently, they can use the caption to do, so the other fields are still redundant.
